cli/main.rs

```diff
@@ -822,3 +833,143 @@ fn wait_for_start(
+  }
+
+  let token = if let Ok(token) = std::env::var("DENO_UNSTABLE_TUNNEL_TOKEN") {
+    token
+  } else {
+    match tools::deploy::get_token_entry()?.get_password() {
+      Ok(token) => token,
+      Err(keyring::Error::NoEntry) => {
+        let token = auth_tunnel().await?;
+
+        if deploy_config.is_none() {
+          // we regenerate the factory & CliOptions since auth_tunnel updates
+          // the config file with the deploy config, only if it was not set previously.
+          factory = CliFactory::from_flags(Arc::new(flags.clone()));
+          cli_options = factory.cli_options()?;
+        }
+
+        token
+      }
+      Err(e) => {
+        init_logging(None, None);
+        exit_for_error(e.into());
+      }
+    }
+  };
+
+  let (org, app) = if let (Ok(org), Ok(app)) = (
+    std::env::var("DENO_UNSTABLE_TUNNEL_ORG"),
+    std::env::var("DENO_UNSTABLE_TUNNEL_APP"),
+  ) {
+    (org, app)
+  } else {
+    let deploy_config = cli_options
+      .start_dir
+      .to_deploy_config()?
+      .expect("auth to be called");
+
+    (deploy_config.org, deploy_config.app)
+  };
+
+  let Some(addr) = tokio::net::lookup_host(&host).await?.next() else {
+    return Ok(());
+  };
+  let Some((hostname, _)) = host.split_once(':') else {
+    return Ok(());
+  };
+
+  let cert_store_provider = factory.root_cert_store_provider();
+  let root_cert_store = cert_store_provider.get_or_try_init()?.clone();
+
+  let (tunnel, metadata, mut events) =
+    match deno_runtime::deno_net::tunnel::TunnelListener::connect(
+      addr,
+      hostname,
+      Some(root_cert_store.clone()),
+      token,
+      org.clone(),
+      app.clone(),
+    )
+    .await
+    {
+      Ok(res) => res,
+      Err(deno_runtime::deno_net::tunnel::Error::InvalidToken) => {
+        tools::deploy::get_token_entry()?.delete_credential()?;
+
+        let token = auth_tunnel().await?;
+        deno_runtime::deno_net::tunnel::TunnelListener::connect(
+          addr,
+          hostname,
+          Some(root_cert_store),
+          token,
+          org,
+          app,
+        )
+        .await?
+      }
+      Err(e) => return Err(e.into()),
+    };
+
+  let addr = tunnel.local_addr()?;
+
+  let endpoint = if addr.port() == 443 {
+    format!("https://{}", addr.hostname())
+  } else {
+    format!("https://{}:{}", addr.hostname(), addr.port())
+  };
+
+  tokio::spawn(async move {
+    while let Some(event) = events.next().await {
+      use deno_runtime::deno_net::tunnel::Event;
+      match event {
+        Event::Routed => {
+          eprintln!(
+            "{}",
+            colors::green(format!("You are connected to {endpoint}!"))
+          );
+        }
+        Event::Migrate => {
+          // TODO: reconnect
+        }
+      }
+    }
+  });
+
+  for (k, v) in metadata.env {
+    // SAFETY: We're doing this before any threads are created.
+    unsafe {
+      std::env::set_var(k, v);
+    }
+  }
+  log::debug!("Endpoint is being routed");
```

Comment:
This will never output because logging isn't initialized. I almost wonder if we should always initialize logging after parsing flags and then have it "elevate" to otel.

Reply:
that might be good, though would prefer to keep that for a followup